### PR TITLE
gh-117641: Use set comprehension for `posixpath.commonpath()`

### PR DESCRIPTION
**Pull request by nineteendo**
### Benchmark

<details><summary>script</summary>

```shell
# test.sh
echo 1 item && python -m timeit -s "import before.posixpath" "before.posixpath.commonpath(['foo'])" && python -m timeit -s "import after.posixpath" "after.posixpath.commonpath(['foo'])"
echo 10 items && python -m timeit -s "import before.posixpath" "before.posixpath.commonpath(['foo'] * 10)" && python -m timeit -s "import after.posixpath" "after.posixpath.commonpath(['foo'] * 10)"
echo 100 items && python -m timeit -s "import before.posixpath" "before.posixpath.commonpath(['foo'] * 100)" && python -m timeit -s "import after.posixpath" "after.posixpath.commonpath(['foo'] * 100)"
```
</details>

```none
1 item
200000 loops, best of 5: 1.81 usec per loop # before
200000 loops, best of 5: 1.47 usec per loop # after
# -> 1.23x faster
10 items
50000 loops, best of 5: 5.25 usec per loop # before
50000 loops, best of 5: 4.66 usec per loop # after
# -> 1.13x faster
100 items
5000 loops, best of 5: 39.9 usec per loop # before
10000 loops, best of 5: 37.3 usec per loop # after
# -> 1.07x faster
```

* Issue: gh-117641

### CODE REVIEW

File: Lib/posixpath.py [hunk 1]
```diff
@@ -572,7 +570,7 @@ def commonpath(paths):
                 common = s1[:i]
                 break
 
-        prefix = sep if isabs else sep[:0]
+        prefix = sep if paths[0][:1] == sep else sep[:0]
```
Comment (eendebakpt) [inline Lib/posixpath.py]: Why not keep `startswith`? Can you also benchmark withs strings that have lengths > 1? The length 1 string '/' you used might have some special behavior.

Comment (nineteendo) [reply Lib/posixpath.py]: ~I can't benchmark `startswith()` on 3.12.~

Comment (nineteendo) [reply Lib/posixpath.py]: Actually, I can just replace the slice comparison in the reference code with `startwith()`. I'll re-run the benchmark with `foo` soon.

File: Lib/posixpath.py [hunk 2]
```diff
@@ -572,7 +570,7 @@ def commonpath(paths):
                 common = s1[:i]
                 break
 
-        prefix = sep if isabs else sep[:0]
+        prefix = sep if paths[0].startswith(sep) else sep[:0]
```
Comment (blhsing) [inline Lib/posixpath.py]: Can `sep[:0]` ever be anything other than `''`?

Comment (eendebakpt) [reply Lib/posixpath.py]: It could be `b''`. But it indeed looks a bit odd

Comment (blhsing) [reply Lib/posixpath.py]: Ah yes got it thanks.

**Comment by nineteendo** [comment]
@serhiy-storchaka, do you have suggestions here?

**Comment by nineteendo** [comment]
I wanted to make the code more readable. But I'll revert it as it's slower:

```zsh
wannes@Stefans-iMac Desktop % python -m timeit -s "import test" "test.test1(['foo'])" && python -m timeit -s "import test" "test.test2(['foo'])"
1000000 loops, best of 5: 369 nsec per loop # unpack
500000 loops, best of 5: 451 nsec per loop # no unpack
# -> 1.22x slower
```

**Comment by nineteendo** [comment]
@erlend-aasland, do you think this can be merged now? I addressed serhiy-storchaka's question.

**Comment by erlend-aasland** [comment]
> @erlend-aasland, do you think this can be merged now? I addressed serhiy-storchaka's question.

I'll run PGO benchmarks for the updated PR first.

**Review by erlend-aasland — APPROVED** [review]

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by nineteendo** [comment]
@erlend-aasland, could you merge it? It has been approved by serhiy.

**Comment by erlend-aasland** [comment]
@nineteendo: yes, it was on my list; there is no need for the extra ping. There is no hurry; I'll land it tomorrow morning. Please avoid unneeded pings.

**Comment by nineteendo** [comment]
Thanks.

> Please avoid unneeded pings.

Sorry, sometimes I get no response, and then I don't know if my message was even read.

**Pull request merged**